### ISSUE TITLE
[v2] AttributeError: 'CacheModule' object has no attribute '_display'

### ISSUE THREAD

**Issue by mgedmin** [ansible/ansible#12708]
**Issue Type**: Bug Report
**Ansible Version**:

```
ansible 2.0.0 (devel eb60ce1b6e) last updated 2015/10/12 09:17:27 (GMT +300)
  lib/ansible/modules/core: (detached HEAD 92d25017ed) last updated 2015/10/12 09:17:45 (GMT +300)
  lib/ansible/modules/extras: (detached HEAD 205115ea1f) last updated 2015/10/12 09:17:45 (GMT +300)
```

**Summary**:
I was testing a custom module, and this happened:

```
$ ANSIBLE_CONFIG=vagrant.cfg ansible-playbook test-changelog.yml 

PLAY ***************************************************************************

TASK [changelog msg=one append=no] *********************************************
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/queues.py", line 264, in _feed
    send(obj)
  File "/home/mg/src/ansible/lib/ansible/vars/__init__.py", line 94, in __getstate__
    fact_cache = self._fact_cache.copy(),
  File "/home/mg/src/ansible/lib/ansible/plugins/cache/__init__.py", line 63, in copy
    return dict(self)
  File "/home/mg/src/ansible/lib/ansible/plugins/cache/__init__.py", line 44, in __getitem__
    return self._plugin.get(key)
  File "/home/mg/src/ansible/lib/ansible/plugins/cache/jsonfile.py", line 72, in get
    self._display.warning("error while trying to write to %s : %s" % (cachefile, str(e)))
AttributeError: 'CacheModule' object has no attribute '_display'
```

At that point ansible-playbook hangs until I hit ^C.

**Steps to Reproduce**:
1. Create an `ansible.cfg`:

``` ini
[defaults]
gathering = smart
fact_caching = jsonfile
fact_caching_connection = .cache/facts/
fact_caching_timeout = 86400
```
1. Create a playbook `test.yml`:

``` yaml
- hosts: localhost
  tasks:
    - debug: msg="hello"
```
1. Create an empty cache file (or any other error condition, e.g. a read-only `.cache` directory)

```
$ mkdir -p .cache/facts
$ echo > .cache/facts/localhost
```
1. Run `ansible-playbook test.yml`

**Expected Results**:

Either the empty cache file is ignored, or there's an error saying that the cache file is corrupt.

**Actual Results**:

```
PLAY ***************************************************************************
Unexpected Exception: 'CacheModule' object has no attribute '_display'
to see the full traceback, use -vvv
```

(there was no hang with this reproducible example)

**Comment by mgedmin** [comment]
For the record, the underlying error is

```
Traceback (most recent call last):
  File "/home/mg/src/ansible/lib/ansible/plugins/cache/jsonfile.py", line 68, in get
    value = json.load(f)
  File "/usr/lib/python2.7/dist-packages/simplejson/__init__.py", line 448, in load
    use_decimal=use_decimal, **kw)
  File "/usr/lib/python2.7/dist-packages/simplejson/__init__.py", line 505, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/dist-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/lib/python2.7/dist-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

because the corresponding cache file (`./cache/facts/precise` is empty).

**Comment by mgedmin** [comment]
I've updated the original description and added steps to reproduce.

**Comment by mgedmin** [comment]
BTW you get an empty cache file when a module returns `ansible_facts` that contain non-string values (such as True):

```
Traceback (most recent call last):
  File "/home/mg/src/ansible/bin/ansible-playbook", line 80, in <module>
    sys.exit(cli.run())
  File "/home/mg/src/ansible/lib/ansible/cli/playbook.py", line 138, in run
    results = pbex.run()
  File "/home/mg/src/ansible/lib/ansible/executor/playbook_executor.py", line 137, in run
    result = self._tqm.run(play=play)
  File "/home/mg/src/ansible/lib/ansible/executor/task_queue_manager.py", line 217, in run
    return strategy.run(iterator, play_context)
  File "/home/mg/src/ansible/lib/ansible/plugins/strategy/linear.py", line 252, in run
    results = self._wait_on_pending_results(iterator)
  File "/home/mg/src/ansible/lib/ansible/plugins/strategy/__init__.py", line 303, in _wait_on_pending_results
    results = self._process_pending_results(iterator)
  File "/home/mg/src/ansible/lib/ansible/plugins/strategy/__init__.py", line 284, in _process_pending_results
    self._variable_manager.set_host_facts(target_host, facts)
  File "/home/mg/src/ansible/lib/ansible/vars/__init__.py", line 550, in set_host_facts
    self._fact_cache.update(host.name, facts)
  File "/home/mg/src/ansible/lib/ansible/plugins/cache/__init__.py", line 75, in update
    self._plugin.set(key, host_cache)
  File "/home/mg/src/ansible/lib/ansible/plugins/cache/jsonfile.py", line 91, in set
    f.write(jsonify(value))
  File "/home/mg/src/ansible/lib/ansible/parsing/utils/jsonify.py", line 42, in jsonify
    return json.dumps(result2, sort_keys=True, indent=indent, ensure_ascii=False)
  File "/usr/lib/python2.7/json/__init__.py", line 250, in dumps
    sort_keys=sort_keys, **kw).encode(obj)
  File "/usr/lib/python2.7/json/encoder.py", line 209, in encode
    chunks = list(chunks)
  File "/usr/lib/python2.7/json/encoder.py", line 434, in _iterencode
    for chunk in _iterencode_dict(o, _current_indent_level):
  File "/usr/lib/python2.7/json/encoder.py", line 390, in _iterencode_dict
    yield _encoder(value)
  File "/usr/lib/python2.7/json/encoder.py", line 39, in encode_basestring
    return '"' + ESCAPE.sub(replace, s) + '"'
TypeError: expected string or buffer
```

I'll file a separate bug.

**Comment by mgedmin** [comment]
I should probably mention that Ansible 1.9.3 shows a traceback about a JSONDecodeError if you end up with a corrupted cache file.

This bug is about the broken error handling code, i.e. AttributeError: for self._display() in Ansible v2.

**Comment by mgedmin** [comment]
git bisect blames commit 851ed45bbf76b74083a0f7e409811418d45ca217 for the broken cache error handling code.

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by mgedmin** [comment]
There's a minor buglet in the fix: an unexpanded `%s` in the error message.

```
PLAY ***************************************************************************
 [WARNING]: error while trying to read .cache/facts//localhost : Expecting value: line 2 column 1 (char 1). Most likely a corrupt file, so erasing and
failing.

ERROR! The JSON cache file %s was corrupt, or did not otherwise contain valid JSON data. It has been removed, so you can re-run your command now.
```

**Comment by mgedmin** [comment]
> There's a minor buglet in the fix: an unexpanded %s in the error message.

Fixed in 426187489f12f6ab594a2dad9818a21c06bc10c2.
